examples/cryptocurrency-advanced/frontend/src/plugins/blockchain.js

```diff
@@ -4,201 +4,140 @@ import * as proto from '../../proto/stubs.js'
-                if (typeof wallet === 'undefined') {
-                  throw new Error('Wallet not found')
-                }
-
-                // get transactions
-                const transactionsMetaData = Exonum.merkleProof(
-                  Exonum.uint8ArrayToHexadecimal(new Uint8Array(wallet.history_hash.data)),
-                  wallet.history_len,
-                  data.wallet_history.proof,
-                  [0, wallet.history_len],
-                  Exonum.Hash
-                )
-
-                if (data.wallet_history.transactions.length !== transactionsMetaData.length) {
-                  // number of transactions in wallet history is not equal
-                  // to number of transactions in array with transactions meta data
-                  throw new Error('Transactions can not be verified')
-                }
-
-                // validate each transaction
-                const transactions = []
-                let index = 0
-
-                for (let transaction of data.wallet_history.transactions) {
-                  const hash = transactionsMetaData[index++]
-                  const buffer = Exonum.hexadecimalToUint8Array(transaction.message)
-                  const bufferWithoutSignature = buffer.subarray(0, buffer.length - 64)
-                  const author = Exonum.uint8ArrayToHexadecimal(buffer.subarray(0, 32))
-                  const signature = Exonum.uint8ArrayToHexadecimal(buffer.subarray(buffer.length - 64, buffer.length));
-
-                  const Transaction = getTransaction(transaction.debug, author)
-
-                  if (Exonum.hash(buffer) !== hash) {
-                    throw new Error('Invalid transaction hash')
-                  }
-
-                  // serialize transaction and compare with message
-                  if (!Transaction.serialize(transaction.debug).every(function (el, i) {
-                    return el === bufferWithoutSignature[i]
-                  })) {
-                    throw new Error('Invalid transaction message')
-                  }
-
-                  if (!Transaction.verifySignature(signature, author, transaction.debug)) {
-                    throw new Error('Invalid transaction signature')
-                  }
-
-                  const transactionData = Object.assign({ hash: hash }, transaction.debug)
-                  if (transactionData.to) {
-                    transactionData.to = Exonum.uint8ArrayToHexadecimal(new Uint8Array(transactionData.to.data))
-                  }
-                  transactions.push(transactionData)
-                }
-
-                return {
-                  block: data.block_proof.block,
-                  wallet: wallet,
-                  transactions: transactions
-                }
-              })
+            .then(({ block_proof, wallet_proof, wallet_history }) => {
+              if (Exonum.verifyBlock(block_proof, validators) !== undefined) throw new Error('Block is invalid')
```

Comment:
Nit / opinion: `if` with the body not separated into a block harms readability (especially after Rust, where blocks are mandatory).

Also, what's going on in this check? `verifyBlock` now throws on failure and returns `void` otherwise.